### PR TITLE
Fuel Gauge API uses syscalls for kernel device driver access

### PR DESCRIPTION
**Pull request by aaronemassey**
Part of https://github.com/zephyrproject-rtos/zephyr/issues/44847 efforts.

This API gave non-syscall access to device drivers. This PR includes a simple renaming the API to be more specific and also adds syscalls.

If requested, I am more than happy to split this up as two separate PRs, one for each commit to keep PRs small.

### CODE REVIEW

**Comment by teburd** [comment]
There needs to be a few things changed here to make these syscalls.

1. A user mode test should be added. In doing so I think you'll run into some failures. The missing z_vrfy_ handlers (see drivers/spi/spi_handlers.c for example).
2. Since the set props takes an array, for usermode (syscalls) you'll need to do some work here to copy the props array into kernel space and verify in the z_vrfy handler. See the verify helpers https://docs.zephyrproject.org/latest/kernel/usermode/syscalls.html#verification-function for some useful macros to deal with that.

**Review by teburd — CHANGES_REQUESTED** [review]
Prior comment meant to be a review and -1 until the test case and vrfy handlers are done

File: include/zephyr/drivers/fuel_gauge.h [hunk 1]
```diff
@@ -182,4 +227,6 @@ __subsystem struct fuel_gauge_driver_api {
 }
 #endif /* __cplusplus */
 
+#include <syscalls/fuel_gauge.h>
```
Comment (dnojiri) [inline include/zephyr/drivers/fuel_gauge.h:230]: This is where 'struct fuel_gauge_get_property' is defined, right? It's referenced already above.

**Comment by aaronemassey** [comment]
Currently depends on https://github.com/zephyrproject-rtos/zephyr/pull/54187

File: include/zephyr/drivers/fuel_gauge.h [hunk 2]
```diff
@@ -149,8 +175,21 @@ struct fuel_gauge_set_property {
  * @return return=0 if successful, return < 0 if getting all properties failed, return > 0 if some
  * properties failed where return=number of failing properties.
  */
-typedef int (*fuel_gauge_get_property_t)(const struct device *dev,
-					 struct fuel_gauge_get_property *props, size_t props_len);
+__syscall int fuel_gauge_get_property(const struct device *dev,
```
[the remark this answers is not in the document]
Comment (aaronemassey) [reply include/zephyr/drivers/fuel_gauge.h:178]: As far as I'm aware, the `struct` part of the struct declaration makes this unambiguous. However, I could see this as confusing especially when doing a regex or tags search for code. I think you're right that a more specific name should be used. I'll keep track of this and do the rename in a separate small PR.

File: include/zephyr/drivers/fuel_gauge.h (hunk 1 above)
Comment (aaronemassey) [reply include/zephyr/drivers/fuel_gauge.h:230]: `<syscalls/fuel_gauge.h>` references code that gets generated later that is not part of this head.

Example:

```
__pinned_func
static inline int fuel_gauge_get_property(const struct device * dev, struct fuel_gauge_get_property * props, size_t props_len)
{
#ifdef CONFIG_USERSPACE
	if (z_syscall_trap()) {
		union { uintptr_t x; const struct device * val; } parm0 = { .val = dev };
		union { uintptr_t x; struct fuel_gauge_get_property * val; } parm1 = { .val = props };
		union { uintptr_t x; size_t val; } parm2 = { .val = props_len };
		return (int) arch_syscall_invoke3(parm0.x, parm1.x, parm2.x, K_SYSCALL_FUEL_GAUGE_GET_PROPERTY);
	}
#endif
	compiler_barrier();
	return z_impl_fuel_gauge_get_property(dev, props, props_len);
}
```

**Comment by aaronemassey** [comment]
> There needs to be a few things changed here to make these syscalls.
> 
> 1. A user mode test should be added. In doing so I think you'll run into some failures. The missing z_vrfy_ handlers (see drivers/spi/spi_handlers.c for example).
> 2. Since the set props takes an array, for usermode (syscalls) you'll need to do some work here to copy the props array into kernel space and verify in the z_vrfy handler. See the verify helpers https://docs.zephyrproject.org/latest/kernel/usermode/syscalls.html#verification-function for some useful macros to deal with that.

Thanks for the resources!

Took me a minute due to a bit of debugging I had to do. It seems we also never used peripheral emulators with zephyr userspace which after some investigation lead to https://github.com/zephyrproject-rtos/zephyr/pull/54187

**Review by teburd — APPROVED** [review]
I don't love the way this test is setup to require an overlay for each platform the test should run on. This is going to be troublesome and inefficient as time moves forward. This is something that should be addressed sooner rather than later. That's not really the focus of this PR, though it does highlight it by adding the arc platform to the test. Arguably every platform that implements userspace with qemu should run this test suite.

The userspace/syscall changes make good sense and look fine to me though.

**Comment by aaronemassey** [comment]
> I don't love the way this test is setup to require an overlay for each platform the test should run on. This is going to be troublesome and inefficient as time moves forward. This is something that should be addressed sooner rather than later. That's not really the focus of this PR, though it does highlight it by adding the arc platform to the test. Arguably every platform that implements userspace with qemu should run this test suite.
> 
> The userspace/syscall changes make good sense and look fine to me though.

That's a good point. I think I can just make a board neutral overlay that gets applied by the `testcase.yaml` file. Will address in a follow-up PR. Thanks, for pointing it out!

File: include/zephyr/drivers/fuel_gauge.h [hunk 3]
```diff
@@ -173,8 +199,21 @@ struct fuel_gauge_set_property {
...
+				  size_t props_len);
+
+static inline int z_impl_fuel_gauge_get_prop(const struct device *dev,
+					     struct fuel_gauge_get_property *props,
+					     size_t props_len)
+{
+	const struct fuel_gauge_driver_api *api = (const struct fuel_gauge_driver_api *)dev->api;
+
+	if (api->get_property == NULL) {
+		return -ENOSYS;
```
Comment (yperess) [inline include/zephyr/drivers/fuel_gauge.h:212]: The errno errors are so odd, I usually use -ENOTSUP, but I'm happy with dealer's choice here.

**Review by yperess — APPROVED** [review]

File: include/zephyr/drivers/fuel_gauge.h (hunk 3 above)
Comment (aaronemassey) [reply include/zephyr/drivers/fuel_gauge.h:212]: I was copying the pattern I saw from the `sensor` and `gpio` driver APIs.

**Pull request merged**